training: big data with skypilot (#131)
tigrisdata/tigris-os-docs: docs/training/big-data-skypilot/index.mdx

[SkyPilot](https://skypilot.readthedocs.io/en/latest/docs/index.html) simplifies
multi-cloud deployments, offering flexibility for training large-scale data
models across different cloud providers. Unlike custom solutions or more
involved tools like Terraform, SkyPilot provides an effortless way to deploy
workloads seamlessly across clouds, eliminating the need for manual
intervention.

While multi-cloud tools like SkyPilot abstract away cloud provider specific
quirks, underlying egress costs and bottlenecks in the storage layer can be
roadblocks. Combining Tigris with SkyPilot ensures there are no cross-cloud
egress costs, all while ensuring consistent performance and reliability in any
region.

In this guide, you'll build a multi-cloud compatible model training job that
leverages Tigris to store training data and SkyPilot to manage compute.

## Prerequisites

- A Tigris account
- [SkyPilot](https://skypilot.readthedocs.io/en/latest/getting-started/installation.html)
  installed
- A bucket you want to use for training and model data (`mybucket`)
- Our
  [demo training repository](https://github.com/tigrisdata-community/skypilot-training-demo)
  cloned

> - Accounts with your desired cloud providers (AWS, Azure, GCP, etc; we'll use

Reviewer: ```suggestion
- Accounts with your desired cloud providers (Lamba Labs, Run Pod, etc; we'll use
```

# Training with Big Data on SkyPilot

[SkyPilot](https://skypilot.readthedocs.io/en/latest/docs/index.html) simplifies

> multi-cloud deployments, offering unmatched flexibility for training large-scale

Reviewer: ```suggestion
multi-cloud deployments, offering flexibility for training large-scale
```

:::

### Importing the model

Once all of the data is imported, the model weights are fetched from Hugging
Face's CDN and then written to Tigris:

```python
model, tokenizer = FastLanguageModel.from_pretrained(
    model_name = model_name, # Qwen/Qwen2.5-0.5B
    max_seq_length = 4096,
    dtype = None, # auto detection
    load_in_4bit = False, # If True, use 4bit quantization to reduce memory usage
)

# Save the model to Tigris
model.save_pretrained(f"{home_dir}/tigris/models/{model_name}")
tokenizer.save_pretrained(f"{home_dir}/tigris/models/{model_name}")
```

Note that the model weights are saved to Tigris using the `{home_dir}/tigris`
folder. Tigris isn't normally a Linux filesystem, but this example uses
[geesefs](https://github.com/yandex-cloud/geesefs) to mount Tigris as a
[FUSE](https://www.kernel.org/doc/html/latest/filesystems/fuse.html) (Filesystem
in Userspace) filesystem. This allows you to use normal filesystem calls to
interact with Tigris, which is useful for AI workloads that assume data is
accessible via a filesystem.

The model has two main components that we care about:

- The model weights themselves, these are the weights that get loaded into GPU

>   ram and used to inference results based on input data.

Reviewer: ```suggestion
  vram and used to inference results based on input data.
```

    },
    {
      "role": "assistant",
      "content": "There are three r's in the word 'strawberry'."
    }
  ]
}
```

And then turns them into pretokenized form like this:

```text
<|im_start|>system
You are a helpful assistant that will make sure that the user gets the correct answer to their question. You are an expert at counting letters in words.
<|im_end|>
<|im_start|>user
Hello, computer, how many r's are in the word 'strawberry'?
<|im_end|>
<|im_start|>assistant
There are three r's in the word 'strawberry'.
<|im_end|>
```

This is what the model actually sees under the hood. Those `<|im_start|>` and
`<|im_end|>` tokens are special tokens that the model and inference runtime use
to know when a new message starts and ends.

### Actually training the model

Once the dataset is pre-chewed and everything else is ready in Tigris, the real
fun can begin. The model is loaded from Tigris:

```python
model = FastLanguageModel.from_pretrained(f"{home_folder}/tigris/models/{model_name}", ...)
```

And then we stack a new [LoRA](https://github.com/microsoft/LoRA) (Low-Rank
Adaptation) model on top of it.

We use a LoRA adapter here because this requires much less system resources to

> train than doing a full-blown finetuing run on the dataset. When you use a LoRA,

Reviewer: ```suggestion
train than doing a full-blown finetuning run on the dataset. When you use a LoRA,
```

  you to use the model for inference right away.

```python
# Save the LoRA model for later use
model.save_pretrained(f"{home_dir}/tigris/done/{model_name}/{dataset_name}/lora_model")
tokenizer.save_pretrained(f"{home_dir}/tigris/done/{model_name}/{dataset_name}/lora_model")

# Save the LoRA fused to the base model for inference with vllm
model.save_pretrained_merged(f"{home_dir}/tigris/done/{model_name}/{dataset_name}/fused", tokenizer, save_method="merged_16bit")
```

### Other things to have in mind

Each of the scripts that run in the instance are designed to be idempotent and
are intended to run in sequence, but `import-dataset.py` and `import-model.py`
can be run in parallel.

Training this LoRA adapter on a dataset of 100k examples takes about 15 minutes
including downloading the dataset, standardizing it, downloading the model,
saving it to Tigris, loading the model, pre-chewing the dataset, and training
the model.

In theory, this example can run on any nVidia GPU with at least 16 GB of vram
and in an environment with the FUSE character device `/dev/fuse` usable. If you
can't use FUSE for some reason, you will need to write custom python code to

> interact with Tigris directly. This is left as an exercise for the reader. Hint:

Reviewer: should this be a guide?

Author: Yes, but right now I'm stubbing it as an exercise for the reader.

use the `multiprocessing` module to parallelize the model weight loading and
saving steps.

## Next steps

Once training is complete, you can use the model weights with
[`sky serve`](https://skypilot.readthedocs.io/en/latest/serving/sky-serve.html):

```text
sky serve up -n mymodel service.yaml --env-file .env
```

It will give you a URL that you can use to interact with your model:

```text
curl 3.84.15.251:30001/v1/chat/completions \
    -X POST \
    -d '{"model": "/home/ubuntu/tigris/done/Qwen/Qwen2.5-0.5B/mlabonne/FineTome-100k/fused", "messages": [{"role": "user", "content": "Who are you?"}]}' \
    -H 'Content-Type: application/json'
```

You can destroy this with `sky serve down`:

```text
sky serve down mymodel

> ```

Reviewer: ```suggestion
```

## What you just built
- A multi-cloud compatible model training job that uses Tigris as intermediate
  and permanent storage for training data and model weights.
- A SkyPilot job that spins up a GPU instance, trains a LoRA adapter on a model,
  and then shuts down the instance after training is complete.

You did great!
```

Once all of the data is imported, the model weights are fetched from Hugging
Face's CDN and then written to Tigris:

```python
model, tokenizer = FastLanguageModel.from_pretrained(
    model_name = model_name,
    max_seq_length = 4096,
    dtype = None, # auto detection
    load_in_4bit = False, # If True, use 4bit quantization to reduce memory usage
)

# Save the model to Tigris
model.save_pretrained(f"{home_dir}/tigris/models/{model_name}")
tokenizer.save_pretrained(f"{home_dir}/tigris/models/{model_name}")
```

You may be asking yourself something like "Wait, why are you saving it to the
disk? I thought it was going to Tigris, Tigris isn't a filesystem, it's object
storage". You're right. Tigris normally is not a filesystem, but S3 (the
technology/API signature that Tigris is built upon) has many semantics that can
somewhat-cleanly-if-you-squint-enough be mapped to the normal Unix filesystem
calls. In order to fit the square peg of Tigris into the round hole of the POSIX
filesystem interface, we use a tool called
[geesefs](https://github.com/yandex-cloud/geesefs). GeeseFS uses the Linux
Kernel's [FUSE](https://www.kernel.org/doc/html/latest/filesystems/fuse.html)

> (Filesystem in USErspace) subsystem to mount an S3 bucket as if it was a

Reviewer 2: `(Filesystem in Userspace)`

Author: This is intentionally spelled out like this so that you can visually see what parts of "FUSE" point to which parts of the acronym. Maybe it's okay without that though.
```suggestion
(Filesystem in Userspace) subsystem to mount an S3 bucket as if it was a
```

filesystem. This allows you to use the normal filesystem calls to interact with
Tigris, which is very useful for AI workloads that assume data is accessible via
a filesystem. We have to do this because the Hugging Face transformers library
assumes that the model weights are on the filesystem, not in object storage.

The model has two main components that we care about:

- The floating-point model weights themselves, these are the weights that get

>   loaded into GPU ram and used to inference results based on input data.

Reviewer 2: `GPU vram`?

Author: ```suggestion
  loaded into GPU vram and used to inference results based on input data.
```
